### ISSUE TITLE
The doc_markdown lint triggers on the link part of reference style links

### ISSUE THREAD

**Issue by jonasbb** [rust-lang/rust-clippy#12075]
### Summary

The doc_markdown lint triggers on the link part of reference style links. If a link contains a `_`, e.g., for rustdoc links`, this should be allowed.

### Lint Name

doc_markdown

### Reproducer

I tried this code:

```rust
#[warn(clippy::doc_markdown)]

/// It is added implicitly by the [`#[serde_as]`] macro when any `schemars` feature is enabled.
///
/// [`#[serde_as]`]: crate::serde_as
struct Foo;
```

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ac33414b276eb1b091f4dfad19cf016e

I saw this happen:

```
warning: item in documentation is missing backticks
 --> src/main.rs:5:22
  |
5 | /// [`#[serde_as]`]: crate::serde_as
  |                      ^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#doc_markdown
note: the lint level is defined here
 --> src/main.rs:1:8
  |
1 | #[warn(clippy::doc_markdown)]
  |        ^^^^^^^^^^^^^^^^^^^^
help: try
  |
5 | /// [`#[serde_as]`]: `crate::serde_as`
  |                      ~~~~~~~~~~~~~~~~~
```

I expected to see this happen:
No error should be shown.

### Version

```text
rustc 1.77.0-nightly (e51e98dde 2023-12-31)
binary: rustc
commit-hash: e51e98dde6a60637b6a71b8105245b629ac3fe77
commit-date: 2023-12-31
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```

**Comment by GuillaumeGomez** [comment]
It's not a false positive. If you want:

```rust
/// [`#[serde_as]`]: crate::serde_as
```

to be considered as a link definition, you need to add an empty line before it. See the rendering of it with and without the empty line [here](https://spec.commonmark.org/dingus/?text=It%20is%20added%20implicitly%20by%20the%20%5B%60stuff%60%5D%20macro%20when%20any%20%60schemars%60%20feature%20is%20enabled.%0A%5B%60stuff%60%5D%3A%20crate%3A%3AFoo%0A%0AIt%20is%20added%20implicitly%20by%20the%20%5B%60stuff%60%5D%20macro%20when%20any%20%60schemars%60%20feature%20is%20enabled.%0A%0A%5B%60stuff%60%5D%3A%20crate%3A%3AFoo).

**Comment by jonasbb** [comment]
This error occurs with the added empty line too. I updated the original issue description and added a link to the playground. My original description was indeed wrong. The problem seems to relate to the closing `]` in the markdown link name.

**Comment by GuillaumeGomez** [comment]
There is still an issue recommending the wrap the link into backticks. Gonna send a fix for it.
